### PR TITLE
refactor push prosecution case to use new MessagePublisher class

### PR DESCRIPTION
**Pull request by arthurashman**
## What

[Link to story](https://dsdmoj.atlassian.net/browse/-LASB570)

Refactors to reuse MessagePublisher class, and allow for CDA to split apart Hearing payloads from Common Platform and send different messages to MAATAPI SQS queue dependent on whether it is prosecution case data or application to court data. 

## Checklist

Before you ask people to review this PR:

- [x] Tests and linters should be passing
- [x] Github should not be reporting conflicts; you should have recently run `git rebase master`.
- [x] Avoid mixing whitespace changes with code changes in the same commit. These make diffs harder to read and conflicts more likely.
- [x] You should have looked at the diff against master and ensured that nothing unexpected is included in your changes.
- [x] You should have checked that the commit messages say why the change was made.

### CODE REVIEW

File: spec/models/maat_api/prosecution_case_spec.rb [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+    it "has a case URN" do
+      expect(prosecution_case.case_urn).to eql("20GD0217100")
+    end
+
+    it "has a jurisdiction type" do
+      expect(prosecution_case.jurisdiction_type).to eql("CROWN")
+    end
+
+    it "has a defendant ASN" do
+      expect(prosecution_case.defendant_asn).to be_nil
```
Comment (laurentqro) [inline spec/models/maat_api/prosecution_case_spec.rb:161]: `has no`

File: app/models/hmcts_common_platform/offence.rb [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    def laa_appln_reference_status_description
+      data.dig(:laaApplnReference, :statusDescription)
+    end
+
+    def laa_appln_reference_laa_contract_number
+      data.dig(:laaApplnReference, :laaContractNumber)
+    end
+
+    def results
+      data[:judicialResults]&.map do |judicial_result_data|
```
Comment (laurentqro) [inline app/models/hmcts_common_platform/offence.rb:62]: `Array(data[:judicialResults]).map...`

File: app/models/maat_api/prosecution_case.rb [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+    def case_creation_date
+      @shared_time.to_date.strftime("%Y-%m-%d")
+    end
+
+    def cjs_location
+      court_centre.short_oucode
+    end
+
+    def doc_language
+      # Should we derive this from somewhere else?
```
Comment (laurentqro) [inline app/models/maat_api/prosecution_case.rb:33]: Indeed, it's worth asking Ahilan if we want to get this with `person_defendant.documentation_language_needs` like we do for Court Applications. Let's make a note of it and do it in a separate PR.

File: app/services/hearings_creator.rb [hunk 4]
```diff
@@ -20,14 +20,24 @@ def call
...
-      prosecution_case[:defendants].each do |defendant|
-        next if defendant[:offences].map { |offence| offence.dig(:laaApplnReference, :applicationReference)&.start_with?("A", "Z") }.any?
-
-        push_to_sqs(shared_time: shared_time,
-                    case_urn: prosecution_case[:prosecutionCaseIdentifier][:caseURN],
-                    defendant: defendant,
-                    appeal_data: nil)
+    hearing[:prosecutionCases]&.each do |prosecution_case_data|
+      prosecution_case_data[:defendants].each do |defendant_data|
+        next if defendant_data[:offences].map { |offence| offence.dig(:laaApplnReference, :applicationReference)&.start_with?("A", "Z") }.any?
```
Comment (laurentqro) [inline app/services/hearings_creator.rb:25]: slightly shorter alternative: `next if defendant_data[:offences].any? { |offence| offence.dig(:laaApplnReference, :applicationReference)&.start_with?("A", "Z") }`

File: app/models/hmcts_common_platform/offence.rb (hunk 2 above)
Comment (arthurashman) [reply app/models/hmcts_common_platform/offence.rb:62]: This then returns an empty array if there are no judicialResults which is not the output we want and would then need to be changed to nil at another point

File: app/services/hearings_creator.rb [hunk 5]
```diff
@@ -20,14 +20,24 @@ def call
...
-        push_to_sqs(shared_time: shared_time,
-                    case_urn: prosecution_case[:prosecutionCaseIdentifier][:caseURN],
-                    defendant: defendant,
-                    appeal_data: nil)
+    hearing[:prosecutionCases]&.each do |prosecution_case_data|
+      prosecution_case_data[:defendants].each do |defendant_data|
+        next if defendant_data[:offences].any? { |offence| offence.dig(:laaApplnReference, :applicationReference)&.start_with?("A", "Z") }
+
+        laa_reference = LaaReference.find_by!(defendant_id: defendant_data[:id], linked: true)
+        next if laa_reference.blank?
```
Comment (laurentqro) [inline app/services/hearings_creator.rb:28]: I don't think we need this. It wasn't in the original implementation, and we now know that all incoming cases are linked.

Comment (arthurashman) [reply app/services/hearings_creator.rb:28]: Quite right. Thank you

**Review by chris-groves — APPROVED** [review]

File: spec/models/maat_api/prosecution_case_spec.rb [hunk 6]
```diff
@@ -43,11 +43,16 @@
   end
 
   it "has no crown_court_outcome" do
```
Comment (laurentqro) [inline spec/models/maat_api/prosecution_case_spec.rb:45]: `it has a crown_court_outcome` :)

File: spec/models/maat_api/prosecution_case_spec.rb [hunk 7]
```diff
@@ -43,11 +43,16 @@
...
+      ccooOutcome: "CONVICTED",
+      caseEndDate: "2021-04-10",
+      appealType: nil,
+    }
+    expect(prosecution_case.crown_court_outcome).to eql(expected)
   end
 
   it "has an inactive" do
-    expect(prosecution_case.inactive).to eql("N")
+    expect(prosecution_case.inactive).to eql("Y")
```
Comment (laurentqro) [inline spec/models/maat_api/prosecution_case_spec.rb:55]: why the change here?

File: spec/models/maat_api/prosecution_case_spec.rb [hunk 8]
```diff
@@ -15,7 +15,7 @@
   end
 
   it "has a jurisdiction type" do
-    expect(prosecution_case.jurisdiction_type).to eql("MAGISTRATES")
+    expect(prosecution_case.jurisdiction_type).to eql("CROWN")
```
Comment (laurentqro) [inline spec/models/maat_api/prosecution_case_spec.rb:18]: why the change here?

Comment (laurentqro) [reply spec/models/maat_api/prosecution_case_spec.rb:18]: ok I get it now :)

File: spec/models/hmcts_common_platform/verdict_spec.rb [hunk 9]
```diff
@@ -0,0 +1,88 @@
+RSpec.describe HmctsCommonPlatform::Verdict, type: :model do
+  let(:data) { JSON.parse(file_fixture("verdict/all_fields.json").read).deep_symbolize_keys }
```
Comment (laurentqro) [inline spec/models/hmcts_common_platform/verdict_spec.rb:2]: should we put this in a context block?

File: spec/models/hmcts_common_platform/defendant_spec.rb [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+    end
+  end
+
+  describe "offences" do
+    it "are HmctsCommonPlatform::Offence objects" do
+      expect(defendant.offences).to all(be_a(HmctsCommonPlatform::Offence))
+      expect(defendant.offences).to be_a(Array)
+    end
+  end
+end
```
Comment (laurentqro) [inline spec/models/hmcts_common_platform/defendant_spec.rb:85]: add a test case for required fields only?

Comment (laurentqro) [reply spec/models/hmcts_common_platform/defendant_spec.rb:85]: I've got this you can use: https://github.com/ministryofjustice/laa-court-data-adaptor/pull/494/files#diff-d1d27e19030c31cfb480424ff6c4cf675d01f5852ca9897c4b57dec2744f637fR1

File: spec/models/maat_api/prosecution_case_spec.rb (hunk 7 above)
Comment (arthurashman) [reply spec/models/maat_api/prosecution_case_spec.rb:55]: Because I have set the default fixture file to being "CROWN" jurisdiction (for CCO) and so inactive calculates to "Y"

**Review by chris-groves — DISMISSED** [review]

**Review by laurentqro — APPROVED** [review]
great work
